### ISSUE TITLE
Mouse is at the bottom on booting on a real hardware

### ISSUE THREAD

**Issue by Vardan2009** [CosmosOS/Cosmos#1695]
The cursor and buttons work fine in VMware.

 But when I boot it on a real hardware. The buttons still work,but the cursor always stays at the bottom and can't go up

**Comment by sayaPintar** [comment]
hey, can you give me the code? please,  i search everytime but doenst work..

**Comment by Vardan2009** [comment]
SYS.MouseManager.ScreenHeight = (uint)canvas.Mode.Rows;
            SYS.MouseManager.ScreenWidth = (uint)canvas.Mode.Columns;

            cursor.x = (int)(canvas.Mode.Rows / 2);
            cursor.y = (int)(canvas.Mode.Columns / 2);

            SYS.MouseManager.X = (uint)(canvas.Mode.Rows / 2);
            SYS.MouseManager.Y = (uint)(canvas.Mode.Columns / 2);


  public bool CheckMousePos()
        {
            cursor.prevx = cursor.x;
            cursor.prevy = cursor.y;
            if ((int)SYS.MouseManager.X != cursor.x || (int)SYS.MouseManager.Y != cursor.y)
            {
                if (SYS.MouseManager.X >= (uint)canvas.Mode.Columns)
                    SYS.MouseManager.X = (uint)canvas.Mode.Columns;
                else if (SYS.MouseManager.Y >= (uint)canvas.Mode.Rows)
                    SYS.MouseManager.Y = (uint)canvas.Mode.Rows;
                cursor.x = (int)SYS.MouseManager.X;
                cursor.y = (int)SYS.MouseManager.Y;
                cursor.changedMouse = true;
            }
            else
                cursor.changedMouse = false;

            if (cursor.changedMouse == true)
                return true;
            //else
            return false;
        }

        private void renderMouse()
        {
            foreach (WindowContainer container in WindowContainers)
            {
                if (container.IsPixelInArray(cursor.prevx, cursor.prevy))
                {
                    //canvas.PutPixel(cursor.prevx, cursor.prevy, container.GetPixelColor(cursor.prevx, cursor.prevy), true);
                    canvas.PutPixel(cursor.prevx, cursor.prevy, container.GetPixelColor(cursor.prevx, cursor.prevy), true);



                }
            }
            canvas.PutPixel(cursor.x, cursor.y, Color.Black, true);

        }


for (; ; ){

                    if (CheckMousePos())
                    {
                        renderMouse();
                    }
}

**Comment by sayaPintar** [comment]
what you Userkit version?

**Comment by sayaPintar** [comment]
btw thanks for the code i Will try it.

**Comment by Vardan2009** [comment]
I use lastest usekit

**Comment by sayaPintar** [comment]
no built in function PutPixel here..

**Comment by sayaPintar** [comment]
you write that?

**Comment by Vardan2009** [comment]
Not all code

**Comment by Vardan2009** [comment]
> no built in function PutPixel here..

Yes. Use canvas.DrawPoint instead

**Comment by sayaPintar** [comment]
how about method IsPixelInArray? sorry for many asking and bad English.

**Comment by quajak** [comment]
Most likely Cosmos doesnt recognise the mouse.

**Comment by Vardan2009** [comment]
@quajak no, the mouse works fine in vmware,but when i boot in on a re hardware,it moves horizontally and stays at the bottom of the screen

**Comment by quajak** [comment]
There are different ways the mouse can be connected. Cosmos does not support USB mice. Do you have the option in the bios to enable legacy ps/2 mouse?

**Comment by napalmtorch** [comment]
I have had the same experience on real hardware. This is due to the PS/2 emuation in the BIOS not being supported by Cosmos. There are a few solutions. Firstly, you can use a PS/2 mouse. Secondly, you could try a different computer. If those options are not available but your keyboard still works, what I did in my OS is made the arrow keys move the mouse and tab to click. Not a perfect solution, but if you NEED a mouse on real hardware this is a good hacky solution

**Comment by Vardan2009** [comment]
I will try to boot on another hardware or try a ps2 mouse. thanks a lot!

**Comment by JuliusCaesar12** [comment]
I'm having this issue with trackpads on two laptops

**Comment by quajak** [comment]
This seems to be a problem in Cosmos, I think there were a few fixes for PS2 in the devkit. Are you using userkit or devkit?

**Comment by JuliusCaesar12** [comment]
Devkit
I just updated it yesterday

**Comment by quajak** [comment]
The only way to fix this would be to debug the `PS2Mouse` driver(look at what packages are being received/why they are being ignored). If you are down to look into this, we can work together on fixing this bug.

**Comment by JuliusCaesar12** [comment]
I exposed the deltaX and deltaY of the IRQ12 incoming packets and found that when trying to move the mouse pointer, the dX value will almost default to -216, and the dY value will almost default to 256. As expected, this shoots the pointer to the bottom-right.

**Comment by quajak** [comment]
Could you please also log `mMouseByte`?

**Comment by JuliusCaesar12** [comment]
Well this is troubling; the BIOS sends the deltas over mMouseByte 0 and 1 instead of 1 and 2

**Comment by quajak** [comment]
This is quite suprising, do you think we can detect when this behavior occurs and then handle the bytes differently?

**Comment by JuliusCaesar12** [comment]
I'm certain that one might be able to (Windows 98 can do it, but Windows 98 doesn't use the scroll wheel normally).
I did find an issue while testing in a virtual machine: moving the scroll wheel will also put the pointer at the bottom.

**Comment by quajak** [comment]
Oh that second bug is interesting, could very well be that the two are connected. Looking at https://wiki.osdev.org/PS/2_Mouse it looks like usually there is not even support for a mouse wheel? Could it be that we are actually getting Intellimouse Z-axis extension packets without intialising it?

**Comment by JuliusCaesar12** [comment]
That is possible, because when I exposed the 'wheel' byte, it was outputting a value.

**Comment by quajak** [comment]
Nice progress! So you were able to break the mouse driver in a virtual machine as well? Your first edit makes it seem like we arnt keeping track of what we are currently receiving correctly. Also how did you disable scrolling?

**Comment by JuliusCaesar12** [comment]
I don't know what I did (I did like four things at once), but now it's working both in a virtual machine and on real hardware. I'm assuming that because the scroll wheel wasn't enabled (which I just commented it out in the initialization method), the mMouseCycle wasn't progressing to three. This is apparently required because otherwise the incoming packets would become out of sync from the operating system since we aren't reading the fourth one (even if there shouldn't be one). I also uncommented the command to set defaults to the mouse and only allowed the the output from the processing of the packet on cycle 3, and if the packets become out-of-sync, the mMouseByte is cleared (I don't know if this is necessary)
